File: PDOO/Material Complementario/Tema7/Villanos-ClaseTeoria/src/villanos/Plan.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package villanos;

import java.util.ArrayList;
import java.util.Date;

/**
 *
 * @author salva
 */
public class Plan {

    private String nombre;
    private Date fechaConsecucion;
    private int prioridad;
    private int posicionPasoSiguiente;
    private boolean terminado;

    //Atributo de referencia
    private ArrayList<Paso> pasos;
    private ObjetivoMaligno miObjetivoMaligno;

    Plan(String nombre, Date fecha, int prioridad) {
        this.nombre = nombre;
        this.fechaConsecucion = fecha;
        this.prioridad = prioridad;
        this.posicionPasoSiguiente = 0;
        this.terminado = false;
        pasos = new ArrayList<>();
        //pasos.setPasos(); //suponiendo que existiera
        miObjetivoMaligno = null;
    }

    /* Opcion para hacer un constructor con parametros opcionales
      Plan(String nombre, Date fecha, int prioridad, int pps, boolean term) {
        this.nombre = nombre;
        this.fechaConsecucion = fecha;
        this.prioridad = prioridad;
        this.posicionPasoSiguiente = pps;
        this.terminado = term;
    }

    Plan (String n, Date f, int p){
        new(n,f,p,0,false);
    }
     */
    public boolean getTerminado() {
    }

    public boolean estaActivo() {
    }

    public void darPasoSiguiente() {
    }

    public int inversionPasoSiguiente() {
    }
}

File: C1/PDOO/Material Complementario/Tema7/VillanosRB-ClaseTeoria/lib/plan.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.
module Villanos
  class Plan
    def initialize(n,f,p, pps = 0, t = false)
      @nombre = n
      @fecha_consecucion = f
      @prioridad = p
      @posicion_paso_siguiente = pps
      @terminado = t
      @pasos = Array.new  # set_pasos # asumiendo que exista
      @objetivo_maligno = nil

    end
    
    def get_terminado
      @terminado
    end
    
    def esta_activo
      
    end
    
    def dar_pasos_siguiente
      paso = @pasos.at(@posicion_paso_siguiente)
      paso.set_terminado(true)
      ganancia = paso.get_ganancia
      if @posicion_paso_siguiente != @pasos.size() -1
        posicion_paso_siguiente +=1
      else
        @terminado = true
      end
      ganancia
    end
    
    def inversion_paso_siguiente
      paso = @pasos.at(@posicion_paso_siguiente)
      gasto = paso.get_inversion
      return gasto
    end
  end
end